### ISSUE TITLE
Salt minion rejected after deleting it from accepted and unaccepted list 

### ISSUE THREAD

**Issue by Meiglus** [saltstack/salt#52982]
### Description of Issue/Question

When I try to start the salt minion on a server, I get the following message. The key was accepted before... 

_Starting The Salt Minion...
Started The Salt Minion.
[CRITICAL] The Salt Master has rejected this minion's public key!
To repair this issue, delete the public key for this minion on the Salt Master and restart this minion. Or restart the Salt Master in open mode to clean out the keys. The Salt Minion will now exit.
salt-minion.service: Main process exited, code=exited, status=77/n/a
salt-minion.service: Failed with result 'exit-code'._

After that I followed some articles how to solve the problem, but couldn't find a solution. I deleted the key with the "salt-key -d" command. Even restarted the whole master. But that didn't change anything. 

At the moment there are ~200 keys registred on the server, and I'm afraid if I delete them all, that won't change anything in the current behaviour. Or the problem suddenly appears on more machines, which would be very unpleasant.

There's no pending key or denied key listed.

### Setup
Master: Ubuntu 16.04 LTS, 2019.2.0
Minion: Ubuntu 18.04 LTS, 2019.2.0
### Steps to Reproduce Issue
Couldn't reproduce so far..

### Versions Report
Salt Version:
           Salt: 2019.2.0

Dependency Versions:
           cffi: Not Installed
       cherrypy: 3.5.0
       dateutil: 2.4.2
      docker-py: Not Installed
          gitdb: 0.6.4
      gitpython: 1.0.1
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 1.0.3
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.12 (default, Nov 12 2018, 14:36:49)
   python-gnupg: 0.3.8
         PyYAML: 3.11
          PyZMQ: 15.2.0
           RAET: Not Installed
          smmap: 0.9.0
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: Ubuntu 16.04 xenial
         locale: UTF-8
        machine: x86_64
        release: 4.4.0-124-generic
         system: Linux
        version: Ubuntu 16.04 xenial

**Comment by garethgreenaway** [comment]
@Meiglus Thanks for the report.  It sounds like maybe there is a conflicting key still on the minion in question.  Can you stop the minion, delete the `pki` directory on the minion in /etc/salt, ensuring the minion's key has been deleted from the master using `salt-key`, and then start the minion back up.  This will ensure that everything is starting from a clean start for this minion.

**Comment by Meiglus** [comment]
Hi @garethgreenaway, thanks for your reply! I did as you told and deleted the pki folder. After that I started the minion and checked the status of it. After 3 seconds it switched to failed with the following message: 

_May 14 08:13:16 servername salt-minion[8424]: [CRITICAL] The Salt Master has rejected this minion's public key!
May 14 08:13:16 servername salt-minion[8424]: To repair this issue, delete the public key for this minion on the Salt Master and restart this minion.
May 14 08:13:16 servername salt-minion[8424]: Or restart the Salt Master in open mode to clean out the keys. The Salt Minion will now exit.
May 14 08:13:16 servername systemd[1]: salt-minion.service: Main process exited, code=exited, status=77/n/a
May 14 08:13:16 servername systemd[1]: salt-minion.service: Failed with result 'exit-code'._

After that I checked the salt-master and the key from the specific minion was in unaccepted keys. So I accepted it, even though the minion-service failed. 

Then I checked again and I got the same error message from the minion.

The key is now listed as accepted in the salt-key -L list. But I can't "talk" to the minion. 

Do you have any idea what I could try?

Thanks for your help!

**Comment by garethgreenaway** [comment]
@Meiglus A couple additional things to check.  On the minion in question, after you've stopped the Salt minion process double check that there isn't another salt-minion process still running.  If there isn't, clear out all the keys as you did before.  Make sure the key ist deleted from the master using `salt-key -d`, wait a second or two and verify it didn't return with `salt-key -L`.  Then start up the minion in debug mode, `salt-minion -l debug`.  You'll see it start up and log some messages about the key being cached.  Once you see that, accept the key on the master.  Then watch the minion logs and try a `test.ping`, see if you see anything that looks like an error.

**Comment by Meiglus** [comment]
@garethgreenaway, I checked as you asked and in our current configuration the salt minion connects to 3 masters. The output said that the minion has a problem to connect to a specific master, so I deleted all the salt masters and just added one. Then I did your cleanup procedure and connected to a single host. 

That worked. 

Then I added the next one, got the error again, cleaned again and could connect. I repeated that three times and now everything works as expected. I can't explain why this worked, but the step-by-step procedure did the job..

Now I don't know if this is a bug or just bad luck, but maybe it can help someone else. :)

Thank you very much for your input!
